Cython/Utility/FunctionArguments.c

```diff
@@ -250,60 +278,427 @@ static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject *const *kwvalues
+        __Pyx_MatchKeywordArg_str(key, argnames, first_kw_arg, index_found, function_name) :
+        __Pyx_MatchKeywordArg_nostr(key, argnames, first_kw_arg, index_found, function_name);
+}
+
+static void __Pyx_RejectUnknownKeyword(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject ** const *first_kw_arg,
+    const char *function_name)
+{
+    // Find the first unknown keyword and raise an error. There must be at least one.
+    Py_ssize_t pos = 0;
+    PyObject *key = NULL;
+
+    while (PyDict_Next(kwds, &pos, &key, NULL)) {
+        // Quickly exclude the 'obviously' valid/known keyword arguments (exact pointer match).
+        PyObject** const *name = first_kw_arg;
+        while (*name && (**name != key)) name++;
+
+        if (!*name) {
+            // No exact match found:
+            // compare against positional (always reject) and keyword (reject unknown) names.
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_INCREF(key);
+            #endif
+
+            size_t index_found = 0;
+            int cmp = __Pyx_MatchKeywordArg(key, argnames, first_kw_arg, &index_found, function_name);
+            if (cmp != 1) {
+                if (cmp == 0) {
+                    PyErr_Format(PyExc_TypeError,
+                        "%s() got an unexpected keyword argument '%U'",
+                        function_name, key);
+                }
+                #if CYTHON_AVOID_BORROWED_REFS
+                Py_DECREF(key);
+                #endif
+
+                break;
+            }
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_DECREF(key);
+            #endif
+        }
+    }
+
+    assert(PyErr_Occurred());
+}
+
+static int __Pyx_ParseKeywordDict(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject *values[],
+    Py_ssize_t num_pos_args,
+    const char* function_name,
+    int ignore_unknown_kwargs)
+{
+    PyObject** const *name;
+    PyObject** const *first_kw_arg = argnames + num_pos_args;
+    Py_ssize_t kwcount, extracted = 0;
+    int failed = 0;
+
+    // Check if dict is unicode-keys-only and let Python set the error otherwise.
+    if (unlikely(!PyArg_ValidateKeywordArguments(kwds))) return -1;
+
+    #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+    kwcount = PyDict_GET_SIZE(kwds);
+    #else
+    kwcount = PyDict_Size(kwds);
+    if (unlikely(kwcount < 0)) return -1;
+    #endif
+
+    // Extract declared keyword arguments.
+    __Pyx_BEGIN_CRITICAL_SECTION(kwds);
+    name = first_kw_arg;
+    while (*name && kwcount > extracted) {
+        PyObject * key = **name;
+        PyObject *value;
+        int found;
+
+        #if __PYX_LIMITED_VERSION_HEX >= 0x030d0000
+        found = PyDict_GetItemRef(kwds, key, &value);
+        #else
+        // There isn't a good way to make this thread-safe before Py3.13.
+        // PyObject_GetItem() would generate lots of expensive KeyErrors.
+        // PyDict_Next() returns borrowed references.
```

Comment:
Should be OK I think. There's nothing between `GetItemWithError` and `INCREF` that can release the GIL so I can't see a thread-safety issue (beyond the usual distrust of borrowed references in PyPy and similar)